File: backend/services/trust_delegation_service.py
"""
Trust Delegation Service
Handles trust-based delegated voting logic
"""
from datetime import datetime, timezone, timedelta
from typing import List, Dict, Any, Optional
import logging

logger = logging.getLogger(__name__)

class TrustDelegationService:
    """Service for managing trust delegations and delegated voting"""
    
    def __init__(self, supabase_client):
        self.supabase = supabase_client
    
    async def create_trust_delegation(
        self, 
        truster_id: str, 
        trusted_id: str, 
        expertise_field: str
    ) -> Dict[str, Any]:
        """
        Create a trust delegation from one user to an expert in a specific field
        """
        try:
            # Check if expert has this expertise
            expertise_check = self.supabase.table('user_expertise')\
                .select('*')\
                .eq('user_id', trusted_id)\
                .eq('expertise_field', expertise_field)\
                .execute()
            
            if not expertise_check.data:
                return {
                    'success': False,
                    'error': 'User is not an expert in this field'
                }
            
            # Create delegation
            delegation = self.supabase.table('trust_delegations').upsert({
                'truster_id': truster_id,
                'trusted_id': trusted_id,
                'expertise_field': expertise_field,
                'is_active': True,
                'revoked_at': None
            }, on_conflict='truster_id,trusted_id,expertise_field').execute()
            
            logger.info(f"Trust delegation created: {truster_id} → {trusted_id} in {expertise_field}")
            
            return {
                'success': True,
                'delegation': delegation.data[0] if delegation.data else None
            }
            
        except Exception as e:
            logger.error(f"Error creating trust delegation: {e}")
            return {'success': False, 'error': str(e)}
    
    async def revoke_trust_delegation(
        self,
        truster_id: str,
        trusted_id: str,
        expertise_field: str
    ) -> Dict[str, Any]:
        """Revoke trust delegation"""
        try:
            result = self.supabase.table('trust_delegations')\
                .update({
                    'is_active': False,
                    'revoked_at': datetime.now(timezone.utc).isoformat()
                })\
                .eq('truster_id', truster_id)\
                .eq('trusted_id', trusted_id)\
                .eq('expertise_field', expertise_field)\
                .execute()
            
            return {'success': True, 'updated': len(result.data) if result.data else 0}
        except Exception as e:
            logger.error(f"Error revoking trust: {e}")
            return {'success': False, 'error': str(e)}
    
    async def trigger_delegated_votes(
        self,
        expert_id: str,
        decision_id: str,
        vote_value: str
    ) -> Dict[str, Any]:
        """
        When expert votes, auto-trigger votes for all users who trust them in this field
        """
        try:
            # Get decision details
            decision = self.supabase.table('demo_decisions')\
                .select('*')\
                .eq('id', decision_id)\
                .single()\
                .execute()
            
            if not decision.data:
                return {'success': False, 'error': 'Decision not found'}
            
            decision_category = decision.data.get('category', 'כללי')
            
            # Get all users who trust this expert in this field
            delegators = self.supabase.table('trust_delegations')\
                .select('truster_id')\
                .eq('trusted_id', expert_id)\
                .eq('expertise_field', decision_category)\
                .eq('is_active', True)\
                .execute()
            
            if not delegators.data:
                logger.info(f"No delegators found for expert {expert_id} in {decision_category}")
                return {'success': True, 'affected_users': 0}
            
            triggered_user_ids = []
            
            # Create delegated votes for each user
            for delegator in delegators.data:
                user_id = delegator['truster_id']
                
                # Check if user already voted directly
                existing_vote = self.supabase.table('user_votes')\
                    .select('*')\
                    .eq('user_id', user_id)\
                    .eq('decision_id', decision_id)\
                    .execute()
                
                if existing_vote.data:
                    # User already voted directly, don't override
                    continue
                
                # Create delegated vote
                self.supabase.table('user_votes').upsert({
                    'user_id': user_id,
                    'decision_id': decision_id,
                    'vote_value': vote_value,
                    'is_delegated': True,
                    'delegated_by': expert_id,
                    'can_withdraw': True,
                    'created_at': datetime.now(timezone.utc).isoformat()
                }, on_conflict='user_id,decision_id').execute()
                
                triggered_user_ids.append(user_id)
                
                # Send notification
                await self.send_delegation_notification(
                    user_id, decision_id, expert_id, vote_value
                )
            
            # Log delegation trigger
            self.supabase.table('vote_delegations_log').insert({
                'decision_id': decision_id,
                'expert_id': expert_id,
                'expert_vote_value': vote_value,
                'affected_users_count': len(triggered_user_ids),
                'votes_triggered': triggered_user_ids
            }).execute()
            
            logger.info(f"Delegated votes triggered: {len(triggered_user_ids)} votes")
            
            return {
                'success': True,
                'affected_users': len(triggered_user_ids),
                'triggered_user_ids': triggered_user_ids
            }
            
        except Exception as e:
            logger.error(f"Error triggering delegated votes: {e}")
            return {'success': False, 'error': str(e)}
    
    async def send_delegation_notification(
        self,
        user_id: str,
        decision_id: str,
        expert_id: str,
        vote_value: str
    ) -> None:
        """Send notification to user about delegated vote"""
        try:
            self.supabase.table('delegation_notifications').insert({
                'user_id': user_id,
                'decision_id': decision_id,
                'expert_id': expert_id,
                'expert_vote': vote_value,
                'notification_type': 'vote_triggered',
                'is_read': False
            }).execute()
        except Exception as e:
            logger.error(f"Error sending notification: {e}")
    
    async def withdraw_delegated_vote(
        self,
        user_id: str,
        decision_id: str,
        action: str,  # 'change' or 'remove'
        new_vote_value: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Allow user to withdraw or change their delegated vote
        Only allowed before withdrawal deadline (3 hours before decision end)
        """
        try:
            # Check if withdrawal is still allowed
            decision = self.supabase.table('demo_decisions')\
                .select('withdrawal_deadline, end_date')\
                .eq('id', decision_id)\
                .single()\
                .execute()
            
            if not decision.data:
                return {'success': False, 'error': 'Decision not found'}
            
            withdrawal_deadline = datetime.fromisoformat(decision.data['withdrawal_deadline'])
            now = datetime.now(timezone.utc)
            
            if now >= withdrawal_deadline:
                return {
                    'success': False, 
                    'error': 'Withdrawal deadline passed',
                    'deadline': withdrawal_deadline.isoformat()
                }
            
            # Get original vote
            original_vote = self.supabase.table('user_votes')\
                .select('*')\
                .eq('user_id', user_id)\
                .eq('decision_id', decision_id)\
                .eq('is_delegated', True)\
                .single()\
                .execute()
            
            if not original_vote.data:
                return {'success': False, 'error': 'No delegated vote found'}
            
            vote_id = original_vote.data['id']
            original_value = original_vote.data['vote_value']
            
            if action == 'remove':
                # Delete the vote
                self.supabase.table('user_votes')\
                    .delete()\
                    .eq('id', vote_id)\
                    .execute()
                
                # Log withdrawal
                self.supabase.table('vote_withdrawals').insert({
                    'vote_id': vote_id,
                    'user_id': user_id,
                    'decision_id': decision_id,
                    'original_vote': original_value,
                    'new_vote': None,
                    'action': 'removed'
                }).execute()
                
                return {'success': True, 'action': 'removed'}
                
            elif action == 'change':
                if not new_vote_value:
                    return {'success': False, 'error': 'New vote value required'}
                
                # Update vote to direct vote
                self.supabase.table('user_votes')\
                    .update({
                        'vote_value': new_vote_value,
                        'is_delegated': False,
                        'delegated_by': None,
                        'can_withdraw': False,
                        'withdrawn_at': datetime.now(timezone.utc).isoformat()
                    })\
                    .eq('id', vote_id)\
                    .execute()
                
                # Log withdrawal
                self.supabase.table('vote_withdrawals').insert({
                    'vote_id': vote_id,
                    'user_id': user_id,
                    'decision_id': decision_id,
                    'original_vote': original_value,
                    'new_vote': new_vote_value,
                    'action': 'changed'
                }).execute()
                
                return {'success': True, 'action': 'changed', 'new_vote': new_vote_value}
            
            return {'success': False, 'error': 'Invalid action'}
            
        except Exception as e:
            logger.error(f"Error withdrawing vote: {e}")
            return {'success': False, 'error': str(e)}
    
    async def get_delegated_votes_for_user(self, user_id: str) -> List[Dict[str, Any]]:
        """Get all active delegated votes for a user with countdown info"""
        try:
            votes = self.supabase.table('user_votes')\
                .select('*, decision:demo_decisions(*), expert:profiles(*)')\
                .eq('user_id', user_id)\
                .eq('is_delegated', True)\
                .eq('can_withdraw', True)\
                .execute()
            
            result = []
            now = datetime.now(timezone.utc)
            
            for vote in votes.data or []:
                decision = vote.get('decision', {})
                withdrawal_deadline = datetime.fromisoformat(decision.get('withdrawal_deadline'))
                time_remaining = withdrawal_deadline - now
                
                result.append({
                    **vote,
                    'time_remaining_seconds': int(time_remaining.total_seconds()),
                    'time_remaining_formatted': self.format_time_remaining(time_remaining),
                    'can_withdraw_now': time_remaining.total_seconds() > 0
                })
            
            return result
            
        except Exception as e:
            logger.error(f"Error getting delegated votes: {e}")
            return []
    
    @staticmethod
    def format_time_remaining(delta: timedelta) -> str:
        """Format time remaining in Hebrew"""
        total_seconds = int(delta.total_seconds())
        
        if total_seconds <= 0:
            return "פג תוקף"
        
        hours = total_seconds // 3600
        minutes = (total_seconds % 3600) // 60
        
        if hours > 0:
            return f"{hours} שעות ו-{minutes} דקות"
        elif minutes > 0:
            return f"{minutes} דקות"
        else:
            return f"{total_seconds} שניות"
    
    async def calculate_vote_results(self, decision_id: str) -> Dict[str, Any]:
        """Calculate vote results including delegation power"""
        try:
            # Get all votes
            votes = self.supabase.table('user_votes')\
                .select('*, delegated_by')\
                .eq('decision_id', decision_id)\
                .execute()
            
            results = {
                'yes': 0,
                'no': 0,
                'abstain': 0,
                'total_votes': 0,
                'direct_votes': 0,
                'delegated_votes': 0,
                'expert_influence': {}
            }
            
            for vote in votes.data or []:
                vote_value = vote['vote_value']
                is_delegated = vote.get('is_delegated', False)
                
                results[vote_value] = results.get(vote_value, 0) + 1
                results['total_votes'] += 1
                
                if is_delegated:
                    results['delegated_votes'] += 1
                    expert_id = vote.get('delegated_by')
                    if expert_id:
                        if expert_id not in results['expert_influence']:
                            results['expert_influence'][expert_id] = 0
                        results['expert_influence'][expert_id] += 1
                else:
                    results['direct_votes'] += 1
            
            return results
            
        except Exception as e:
            logger.error(f"Error calculating results: {e}")
            return {'error': str(e)}

File: frontend/src/services/trustDelegationService.ts
/**
 * Trust Delegation Service
 * Frontend service for trust-based delegated voting
 */

import { supabase } from '@/integrations/supabase/client';

export interface TrustDelegation {
  id: string;
  truster_id: string;
  trusted_id: string;
  expertise_field: string;
  is_active: boolean;
  created_at: string;
  revoked_at?: string;
}

export interface DelegatedVote {
  id: string;
  user_id: string;
  decision_id: string;
  vote_value: string;
  is_delegated: boolean;
  delegated_by?: string;
  can_withdraw: boolean;
  created_at: string;
  withdrawn_at?: string;
  decision?: any;
  time_remaining_seconds?: number;
  can_withdraw_now?: boolean;
}

export const trustDelegationService = {
  /**
   * Create trust delegation to an expert in a specific field
   */
  async createDelegation(trusterId: string, trustedId: string, expertiseField: string) {
    try {
      const { data, error } = await supabase
        .from('trust_delegations')
        .upsert({
          truster_id: trusterId,
          trusted_id: trustedId,
          expertise_field: expertiseField,
          is_active: true
        }, { onConflict: 'truster_id,trusted_id,expertise_field' })
        .select()
        .single();

      if (error) throw error;

      return { success: true, delegation: data };
    } catch (error) {
      console.error('Error creating delegation:', error);
      return { success: false, error };
    }
  },

  /**
   * Revoke trust delegation
   */
  async revokeDelegation(trusterId: string, trustedId: string, expertiseField: string) {
    try {
      const { data, error } = await supabase
        .from('trust_delegations')
        .update({
          is_active: false,
          revoked_at: new Date().toISOString()
        })
        .eq('truster_id', trusterId)
        .eq('trusted_id', trustedId)
        .eq('expertise_field', expertiseField)
        .select();

      if (error) throw error;

      return { success: true, data };
    } catch (error) {
      console.error('Error revoking delegation:', error);
      return { success: false, error };
    }
  },

  /**
   * Expert votes and triggers delegated votes
   */
  async expertVote(expertId: string, decisionId: string, voteValue: string) {
    try {
      // 1. Record expert's vote
      const { data: expertVote, error: voteError } = await supabase
        .from('user_votes')
        .upsert({
          user_id: expertId,
          decision_id: decisionId,
          vote_value: voteValue,
          is_delegated: false
        }, { onConflict: 'user_id,decision_id' })
        .select()
        .single();

      if (voteError) throw voteError;

      // 2. Get decision details
      const { data: decision, error: decisionError } = await supabase
        .from('demo_decisions')
        .select('category, withdrawal_deadline')
        .eq('id', decisionId)
        .single();

      if (decisionError) throw decisionError;

      const category = decision?.category || 'כללי';

      // 3. Get all users who trust this expert in this field
      const { data: delegators, error: delegatorsError } = await supabase
        .from('trust_delegations')
        .select('truster_id')
        .eq('trusted_id', expertId)
        .eq('expertise_field', category)
        .eq('is_active', true);

      if (delegatorsError) throw delegatorsError;

      const triggeredVotes = [];

      // 4. Create delegated votes
      for (const delegator of (delegators || [])) {
        const userId = delegator.truster_id;

        // Skip if user already voted directly
        const { data: existing } = await supabase
          .from('user_votes')
          .select('is_delegated')
          .eq('user_id', userId)
          .eq('decision_id', decisionId)
          .maybeSingle();

        if (existing && !existing.is_delegated) {
          continue; // User voted directly
        }

        // Create delegated vote
        await supabase.from('user_votes').upsert({
          user_id: userId,
          decision_id: decisionId,
          vote_value: voteValue,
          is_delegated: true,
          delegated_by: expertId,
          can_withdraw: true
        }, { onConflict: 'user_id,decision_id' });

        triggeredVotes.push(userId);

        // Send notification
        await supabase.from('delegation_notifications').insert({
          user_id: userId,
          decision_id: decisionId,
          expert_id: expertId,
          expert_vote: voteValue,
          notification_type: 'vote_triggered'
        });
      }

      // 5. Log delegation
      await supabase.from('vote_delegations_log').insert({
        decision_id: decisionId,
        expert_id: expertId,
        expert_vote_value: voteValue,
        affected_users_count: triggeredVotes.length,
        votes_triggered: triggeredVotes
      });

      return {
        success: true,
        expertVote,
        delegatedVotesTriggered: triggeredVotes.length,
        affectedUsers: triggeredVotes
      };
    } catch (error) {
      console.error('Error in expert vote:', error);
      return { success: false, error };
    }
  },

  /**
   * Withdraw or change delegated vote
   */
  async withdrawVote(
    userId: string,
    decisionId: string,
    action: 'change' | 'remove',
    newVoteValue?: string
  ) {
    try {
      // Check deadline
      const { data: decision } = await supabase
        .from('demo_decisions')
        .select('withdrawal_deadline')
        .eq('id', decisionId)
        .single();

      if (!decision) {
        return { success: false, error: 'Decision not found' };
      }

      const deadline = new Date(decision.withdrawal_deadline);
      if (new Date() >= deadline) {
        return { success: false, error: 'Withdrawal deadline has passed' };
      }

      // Get vote
      const { data: vote } = await supabase
        .from('user_votes')
        .select('*')
        .eq('user_id', userId)
        .eq('decision_id', decisionId)
        .single();

      if (!vote) {
        return { success: false, error: 'Vote not found' };
      }

      const originalValue = vote.vote_value;

      if (action === 'remove') {
        // Delete vote
        await supabase.from('user_votes').delete().eq('id', vote.id);

        // Log withdrawal
        await supabase.from('vote_withdrawals').insert({
          vote_id: vote.id,
          user_id: userId,
          decision_id: decisionId,
          original_vote: originalValue,
          action: 'removed'
        });

        return { success: true, action: 'removed' };
      } else if (action === 'change') {
        // Update to direct vote
        await supabase
          .from('user_votes')
          .update({
            vote_value: newVoteValue,
            is_delegated: false,
            delegated_by: null,
            can_withdraw: false,
            withdrawn_at: new Date().toISOString()
          })
          .eq('id', vote.id);

        // Log withdrawal
        await supabase.from('vote_withdrawals').insert({
          vote_id: vote.id,
          user_id: userId,
          decision_id: decisionId,
          original_vote: originalValue,
          new_vote: newVoteValue,
          action: 'changed'
        });

        return { success: true, action: 'changed', newVote: newVoteValue };
      }

      return { success: false, error: 'Invalid action' };
    } catch (error) {
      console.error('Error withdrawing vote:', error);
      return { success: false, error };
    }
  },

  /**
   * Get user's active delegated votes with countdown info
   */
  async getMyDelegatedVotes(userId: string): Promise<DelegatedVote[]> {
    try {
      const { data, error } = await supabase
        .from('user_votes')
        .select(`
          *,
          decision:demo_decisions(*),
          expert:profiles!user_votes_delegated_by_fkey(*)
        `)
        .eq('user_id', userId)
        .eq('is_delegated', true)
        .eq('can_withdraw', true);

      if (error) throw error;

      const now = new Date();
      
      return (data || []).map(vote => {
        const decision = vote.decision;
        const deadline = new Date(decision?.withdrawal_deadline);
        const timeRemaining = Math.floor((deadline.getTime() - now.getTime()) / 1000);

        return {
          ...vote,
          time_remaining_seconds: Math.max(0, timeRemaining),
          can_withdraw_now: timeRemaining > 0
        };
      });
    } catch (error) {
      console.error('Error getting delegated votes:', error);
      return [];
    }
  },

  /**
   * Calculate time remaining formatted
   */
  formatTimeRemaining(seconds: number): string {
    if (seconds <= 0) return 'פג תוקף';

    const hours = Math.floor(seconds / 3600);
    const minutes = Math.floor((seconds % 3600) / 60);

    if (hours > 0) {
      return `${hours} שעות ו-${minutes} דקות`;
    } else if (minutes > 0) {
      return `${minutes} דקות`;
    } else {
      return `${seconds} שניות`;
    }
  }
};
